packages/extension/src/ui/states/tokens.ts

```diff
@@ -1,128 +1,124 @@
-        }))
-      },
-      removeToken: (token: TokenDetails) => {
-        set((state) => ({
-          tokens: state.tokens.filter((t) => t.address !== token.address),
+const mapTokenToTokenDetails = (token: Token): TokenDetails => ({
+  ...token,
+  decimals: number.toBN(token.decimals),
+})
+
+const mapTokenDetailsToToken = (token: Required<TokenDetails>): Token => ({
+  ...token,
+  decimals: token.decimals.toString(),
+})
+
+export const useTokens = create<State>((set, get) => ({
+  tokens: [],
+  addToken: async (token: Required<TokenDetails>) => {
+    if (!isValidAddress(token.address)) {
+      throw Error("token address malformed")
+    }
+    const equalTokenHit = get().tokens.find((t) => equalToken(t, token))
+    if (equalTokenHit) {
+      // if token already exists, but was hidden without balance, show it
+      if (!equalTokenHit.showAlways) {
+        return set((state) => ({
+          ...state,
+          tokens: state.tokens.map((t) =>
+            equalToken(t, equalTokenHit) ? { ...t, showAlways: true } : t,
+          ),
         }))
-      },
-    }),
-    {
-      name: "tokens", // name of item in the storage (must be unique)
-      getStorage: () => localStorage, // (optional) by default the 'localStorage' is used
-      merge: (persistedState, currentState) => ({
-        ...currentState,
-        ...persistedState,
-        tokens: [...currentState.tokens, ...persistedState.tokens],
-      }),
-      partialize: (state) =>
-        Object.fromEntries(
-          Object.entries(state).map(([key, value]) => {
-            if (key === "tokens") {
-              return [
-                key,
-                (value as TokenDetails[]).filter(
-                  (token) =>
-                    !parsedDefaultTokens.some((defaultToken) =>
-                      equalToken(token, defaultToken),
-                    ),
-                ),
-              ]
-            }
-            return [key, value]
-          }),
-        ),
-      deserialize: (str) =>
-        JSON.parse(str, (_, v) => {
-          if (
-            typeof v === "object" &&
-            "type" in v &&
-            "hex" in v &&
-            v.type === "BigNumber"
-          ) {
-            return BigNumber.from(v.hex)
-          }
-          return v
-        }),
-    },
-  ),
-)
+      }
+      throw Error("token already added")
+    }
+    const newToken = { ...token, showAlways: true }
+    await addTokenMsg(mapTokenDetailsToToken(newToken))
+
+    // optimistic update
+    set((state) => ({
+      tokens: [...state.tokens, newToken],
+    }))
+  },
+  removeToken: async (tokenAddress: TokenDetails["address"]) => {
+    await removeTokenMsg(tokenAddress)
+
+    // optimistic update
+    set((state) => ({
+      tokens: state.tokens.filter((t) => t.address !== tokenAddress),
+    }))
+  },
+}))
+
+export const useTokensSubscription = () => {
+  const { data: tokens = [] } = useSWRImmutable("tokens", () => getTokens(), {
```

Comment:
if `getTokens` doesn't use the `this` keyword, this is better written as

```suggestion
  const { data: tokens = [] } = useSWRImmutable("tokens", getTokens, {
```